File: mirror/tier2-ruby/showcase/pattern_nested.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Pattern Matching
# Module: pattern_nested.rb
# Purpose: Demonstrate nested pattern matching

def match_nested_hash(data)
  if data.is_a?(Hash)
    if data["user"].is_a?(Hash)
      user = data["user"]
      if user["profile"].is_a?(Hash)
        profile = user["profile"]
        name = profile["name"] || "unknown"
        {found: true, name: name}
      end
    end
  end
  {found: false}
end

def match_nested_list(data)
  if data.is_a?(Array) && data.size > 0
    first = data[0]
    if first.is_a?(Array) && first.size > 0
      inner = first[0]
      {outer: first, inner: inner}
    end
  end
end

def traverse_structure(obj, depth = 0)
  return nil if depth > 3
  
  case obj
  when Hash
    obj.each { |k, v| traverse_structure(v, depth + 1) if v.is_a?(Hash) || v.is_a?(Array) }
    "traversed_hash"
  when Array
    obj.each { |item| traverse_structure(item, depth + 1) if item.is_a?(Hash) || item.is_a?(Array) }
    "traversed_list"
  end
end

result = {
  nested_hash_found: match_nested_hash({"user" => {"profile" => {"name" => "Alice"}}}),
  nested_hash_notfound: match_nested_hash({"data" => "simple"}),
  nested_list: match_nested_list([[1, 2, 3], 4, 5]),
  traversed: traverse_structure({"a" => {"b" => {"c" => 1}}}),
}

puts result.inspect

File: mirror/tier2-dart/showcase/pattern_nested.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Pattern Matching
// Module: pattern_nested.dart

Map<String, dynamic> matchNestedHash(Map<String, dynamic>? data) {
  if (data == null) return {'found': false};
  
  final user = data['user'];
  if (user is Map) {
    final profile = user['profile'];
    if (profile is Map) {
      final name = profile['name'] ?? 'unknown';
      return {'found': true, 'name': name};
    }
  }
  return {'found': false};
}

Map<String, dynamic>? matchNestedList(List<dynamic> data) {
  if (data.isNotEmpty && data[0] is List) {
    final first = data[0] as List;
    if (first.isNotEmpty) {
      return {'outer': first, 'inner': first[0]};
    }
  }
  return null;
}

void traverseStructure(dynamic obj, {int depth = 0}) {
  if (depth > 3) return;
  
  if (obj is Map) {
    obj.forEach((k, v) {
      if (v is Map || v is List) {
        traverseStructure(v, depth: depth + 1);
      }
    });
  } else if (obj is List) {
    for (final item in obj) {
      if (item is Map || item is List) {
        traverseStructure(item, depth: depth + 1);
      }
    }
  }
}

void main() {
  final result = <String, dynamic>{
    'nestedHashFound': matchNestedHash({'user': {'profile': {'name': 'Alice'}}}),
    'nestedHashNotfound': matchNestedHash({'data': 'simple'}),
    'nestedList': matchNestedList([[1, 2, 3], 4, 5]),
    'traversed': 'done',
  };
  print(result);
}
